File: src/types/PersonalityTrait.ts
import * as DB from "tsondb/schema/dsl"
import { NestedTranslationMap } from "./Locale.js"
import { PersonalityTraitIdentifier } from "./_Identifier.js"
import { PersonalityTraitPrerequisites } from "./_Prerequisite.js"
import { Errata } from "./source/_Erratum.js"
import { src } from "./source/_PublicationRef.js"

export const PersonalityTrait = DB.Entity(import.meta.url, {
  name: "PersonalityTrait",
  namePlural: "PersonalityTraits",
  comment:
    "A personality trait describes character aspects of a person from a certain region. Higher trait levels only cover a part of the region covered by lower-level traits.",
  type: () =>
    DB.Object({
      level: DB.Required({
        comment: "The personality trait’s level.",
        type: DB.Integer({ minimum: 1, maximum: 3 }),
      }),
      prerequisites: DB.Optional({
        type: DB.IncludeIdentifier(PersonalityTraitPrerequisites),
      }),
      combination_options: DB.Optional({
        comment: "The lower-level personality trait(s) this trait can be combined with.",
        type: DB.Array(PersonalityTraitIdentifier(), { minItems: 1, uniqueItems: true }),
      }),
      src,
      translations: NestedTranslationMap(
        DB.Required,
        "PersonalityTrait",
        DB.Object({
          name: DB.Required({
            comment: "The personality trait’s name.",
            type: DB.String({ minLength: 1 }),
          }),
          effects: DB.Required({
            comment:
              "The effects of the personality trait. They should be sorted like they are in the book.",
            type: DB.Array(DB.IncludeIdentifier(PersonalityTraitEffect), { minItems: 1 }),
          }),
          errata: DB.Optional({
            type: DB.IncludeIdentifier(Errata),
          }),
        }),
      ),
    }),
  instanceDisplayName: {},
  uniqueConstraints: [
    {
      entityMapKeyPath: "translations",
      keyPathInEntityMap: "name",
    },
  ],
})

const PersonalityTraitEffect = DB.TypeAlias(import.meta.url, {
  name: "PersonalityTraitEffect",
  type: () =>
    DB.Object({
      label: DB.Optional({
        comment: "A label that is displayed and placed before the actual text.",
        type: DB.String({ minLength: 1 }),
      }),
      text: DB.Required({
        comment: "The effect text.",
        type: DB.String({ minLength: 1 }),
      }),
    }),
})

File: Sources/OptolithDatabaseSchema/Types/PersonalityTrait.swift
import FileDB

/// A personality trait describes character aspects of a person from a certain region. Higher trait levels only cover a part of the region covered by lower-level traits.
@Model
public struct PersonalityTrait {
    /// The personality trait’s level.
    @Minimum(1)
    @Maximum(3)
    let level: Int

    let prerequisites: PersonalityTraitPrerequisites?

    /// The lower-level personality trait(s) this trait can be combined with.
    @MinItems(1)
    @UniqueItems
    @Relationship(PersonalityTrait.self)
    let combination_options: [PersonalityTrait.ID]?

    /// The publications where you can find the entry.
    @MinItems(1)
    let src: [PublicationRef]

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [Locale.ID: Translation]

    @Embedded
    struct Translation {  // PersonalityTraitTranslation
        /// The personality trait’s name.
        @MinLength(1)
        let name: String
        /// The effects of the personality trait. They should be sorted like they are in the book.
        @MinItems(1)
        let effects: [PersonalityTraitEffect]

        /// A list of errata for the entry in the specific language.
        @MinItems(1)
        let errata: [Erratum]?
    }
}

@Embedded
public struct PersonalityTraitEffect {
    /// A label that is displayed and placed before the actual text.
    @MinLength(1)
    let label: String?

    /// The effect text.
    @MinLength(1)
    let text: String
}
